feat(Vehicle): Added option to override vehicle weapons.

George0828Zhang opened the pull request:
The special ammo option did not support vehicle weapons. This commit adds a dedicated option for custom vehicle weapons.

George0828Zhang committed:
feat(Vehicle): Added option to override vehicle weapons.

The special ammo option did not support vehicle weapons. This commit
adds a dedicated option for custom vehicle weapons, with additional
controls such as speed and range.

George0828Zhang commented on src/backend/looped/vehicle/vehicle_ammo_special_type.cpp:
```diff
@@ -0,0 +1,216 @@
+				p_modified_rocket_weapon->m_reload_time_sp      = modified_rocket_reload_time_sp;
+				p_modified_rocket_weapon                        = nullptr;
+
+				if (p_modified_rocket_ammo == nullptr)
```
Only did it to be safe, but you're right it'll probably never be null.

George0828Zhang commented on src/views/vehicle/view_vehicle.cpp:
```diff
@@ -89,6 +91,89 @@ namespace big
 			components::command_checkbox<"vehiclecontrol">();
 			components::command_checkbox<"keepvehicleclean">();
 
+			ImGui::Checkbox("Custom Ammo", &g.vehicle.vehicle_ammo_special.toggle);
```
I was struggling between self->weapons, vehicle->general, or vehicle->weapons(new tab). I agree now that fun options is a good place for it.

George0828Zhang committed:
refactor(Vehicle Weapons): revise implementation based on suggestions.

1. Refactor into a looped_command class.
2. Moved view to Vehicle->fun options.
3. Fixed tank turret and mortar not working when custom weapon enabled.
4. Added missile range, time b4 homing, and tool tips.

George0828Zhang commented on src/backend/looped/vehicle/vehicle_ammo_special_type.cpp:
```diff
@@ -0,0 +1,216 @@
+	static CAmmoRocketInfo::sExplosion modified_rocket_explosion{};
+	static float modified_rocket_lifetime           = 5;
+	static float modified_rocket_launch_speed       = 1250;
+	static CHomingRocketParams modified_rocket_homing_params{};
```
HI, I have moved them into a looped command class, could you check if this is what you're looking for? Thanks.

Yimura commented on src/backend/looped/vehicle/vehicle_ammo_special_type.cpp:
```diff
@@ -0,0 +1,229 @@
+				apply_rocket(g.vehicle.vehicle_ammo_special);
+			}
+		}
+		virtual void on_disable() override
```
Newline between member functions.

George0828Zhang replied:
Done!

Yimura commented on src/backend/looped/vehicle/vehicle_ammo_special_type.cpp:
```diff
@@ -0,0 +1,229 @@
+			restore_mg();
+			restore_rocket();
+		}
+		bool is_weapon_mg(const CWeaponInfo* weapon_info)
+		{
```
Newline between member functions.

Yimura commented on src/backend/looped/vehicle/vehicle_ammo_special_type.cpp:
```diff
@@ -0,0 +1,229 @@
+		{
+			return weapon_info->m_fire_type == eFireType::InstantHit || weapon_info->m_fire_type == eFireType::DelayedHit;
+		}
+		bool is_weapon_rocket(const CWeaponInfo* weapon_info)
```
Newline between member functions.

Yimura commented on src/backend/looped/vehicle/vehicle_ammo_special_type.cpp:
```diff
@@ -0,0 +1,229 @@
+			uint32_t trail_hash          = rocket_info->m_m_trail_fx_hash;
+			return (trail_hash == RAGE_JOAAT("proj_rpg_trail") || trail_hash == RAGE_JOAAT("proj_xm_thruster_rpg_trail"));
+		}
+		void restore_mg()
```
Newline between member functions.

Yimura commented on src/backend/looped/vehicle/vehicle_ammo_special_type.cpp:
```diff
@@ -0,0 +1,229 @@
+				m_mg_weapon_info->m_weapon_range                   = m_mg_range;
+			}
+		}
+		void backup_mg(CWeaponInfo* weapon_info)
```
Newline between member functions.

Yimura commented on src/backend/looped/vehicle/vehicle_ammo_special_type.cpp:
```diff
@@ -0,0 +1,229 @@
+			m_mg_weapon_info->m_alternate_wait_time            = g_vehicle_ammo_settings.alternate_wait_time;
+			m_mg_weapon_info->m_weapon_range                   = g_vehicle_ammo_settings.weapon_range;
+		}
+		void restore_rocket()
```
Newline between member functions.

Yimura commented on src/backend/looped/vehicle/vehicle_ammo_special_type.cpp:
```diff
@@ -0,0 +1,229 @@
+				m_rocket_weapon_info                = nullptr;
+			}
+		}
+		void backup_rocket(CWeaponInfo* weapon_info)
```
Newline between member functions.

Yimura commented on src/backend/looped/vehicle/vehicle_ammo_special_type.cpp:
```diff
@@ -0,0 +1,229 @@
+			m_rocket_time_before_homing  = rocket_info->m_time_before_homing;
+			m_rocket_homing_params       = rocket_info->m_homing_rocket_params;
+		}
+		void apply_rocket(const vehicle_ammo_setting& g_vehicle_ammo_settings)
```
Newline between member functions.

Yimura reviewed (changes requested):
Very nicely structured code, this must've taken a bunch of work to get working.

Just a few nitpicks really.

George0828Zhang committed:
chore(Vehicle weapons): Newline between functions and update tooltip.

1. Add newline as suggested.
2. Tooltip for "Time Before Homing" updated to reflect correct purpose.

Yimura reviewed (dismissed):
I don't have any complaints with it other than that it doesn't make much use of translation keys.

@George0828Zhang would you be willing to open up a PR on https://github.com/YimMenu/Translations to add these keys? Don't forget to bump the version in `index.json`.

George0828Zhang committed:
fix(Vehicle Weapons): Use translation keys. Further improves missiles.

George0828Zhang commented on src/backend/looped/vehicle/vehicle_ammo_special_type.cpp:
```diff
@@ -0,0 +1,248 @@
+
+		void backup_rocket(CWeaponInfo* weapon_info)
+		{
+			m_rocket_weapon_info         = weapon_info;
```
It is impossible because 
1. weapon_info is guaranteed != null
2. if m_mg_weapon_info is null, then [m_mg_weapon_info != weapon_info](https://github.com/George0828Zhang/YimMenu/blob/f1c2dfdd6836e4ac8695d9dda8c9fe516c6829d6/src/backend/looped/vehicle/vehicle_ammo_special_type.cpp#L49), meaning unarmed -> equipped weapon
3. backup_mg(weapon_info) would've been called, setting m_mg_weapon_info to valid value.
The null check in restore() is required, otherwise error arise when you toggle off without ever changing weapons.

George0828Zhang commented on src/backend/looped/vehicle/vehicle_ammo_special_type.cpp:
```diff
@@ -0,0 +1,248 @@
+{
+	class custom_vehicle_weapon : looped_command
+	{
+		using looped_command::looped_command;
```
It's leftover from `src/backend/looped/vehicle/fly.cpp` which I based this on. Why is it an issue though?

George0828Zhang replied:
It's not like I left 100 debug functions in there, it's 1 line. Maybe I can make changes in batch if more issues arise.

Yimura replied:
@gir489returns all command implementations do this.

Yimura committed:
Merge branch 'master' into vehicle_weapons

Yimura reviewed (approved):
Translations have been implemented so on my part it's good.
